### PR TITLE
[WIP] Initial design of URL Rewrite

### PR DESCRIPTION
**Pull request by ZestyBread**
Big time WIP. Initial design based on Damian and David's meeting. 
- Supports loading rules from a config file (current project uses JSON)
- Supports URL path rewrites and redirects
- Supports URL scheme rewrites and redirects
- Follows the builder model

Items to consider/work on in future design:
- How to handle rewriting whole URL? Specifying specific components vs regex on whole thing
- Use routing as a basis for matching and rewriting?
- Redirect status codes, handling?
- Absolute vs path and query for URLs?
- General design.

### CODE REVIEW

**Comment by Tratcher** [comment]
@DamianEdwards @davidfowl @muratg

File: samples/Rewrite/Startup.cs [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+        }
+
+        public IConfiguration Configuration { get; set; }
+        // This method gets called by the runtime. Use this method to configure the HTTP request pipeline.
+        public void Configure(IApplicationBuilder app)
+        {
+
+            var rewriteBuilder = new UrlRewriteBuilder();
+            rewriteBuilder.RulesFromConfig(Configuration);
+            app.UseRewriter(rewriteBuilder.Build());
```
Comment (Tratcher) [inline samples/Rewrite/Startup.cs]: I bet we'll end up with an extension like `app.UseRewriter(Configuration);` that hides this.

File: test/RewriteTest/Rewrite2MiddlewareTests.cs [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+            var builder = new ApplicationBuilder(serviceProvider: null);
+            var rewriteBuilder = new UrlRewriteBuilder();
+            rewriteBuilder.RedirectScheme(30);
+            builder.UseRewriter(rewriteBuilder.Build());
+            var app = builder.Build();
+            app.Invoke(context).Wait();
+            Assert.True(context.Response.Headers["location"].First().StartsWith("https"));
+        }
+
+        [Theory]
```
Comment (Tratcher) [inline test/RewriteTest/Rewrite2MiddlewareTests.cs]: `[Theory(SkipReason="Not Implemented")]`

File: samples/Rewrite/Structure2/UrlRewriteMiddleware.cs [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+                    if (rule.RuleState == Transformation.Redirect)
+                    {
+                        return;
+                    }
+                    else if (rule.RuleState == Transformation.TerminatingRewrite)
+                    {
+                        break;
+                    }
+                }
+                // TODO check type of rule and go from there, either check class type or a private bool
```
Comment (Tratcher) [inline samples/Rewrite/Structure2/UrlRewriteMiddleware.cs]: This is done, right?

File: samples/Rewrite/FileParser/RewriteConfigurationExtensions.cs [hunk 4]
```diff
@@ -0,0 +1,74 @@
+﻿using Microsoft.Extensions.Configuration;
```
Comment (muratg) [inline samples/Rewrite/FileParser/RewriteConfigurationExtensions.cs]: Copyright headers (here and on the other files)

File: samples/Rewrite/FileParser/RewriteConfigurationExtensions.cs [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+        /// <param name="reloadOnChange">Whether the configuration should be reloaded if the file changes.</param>
+        /// <returns>The <see cref="IConfigurationBuilder"/>.</returns>
+        public static IConfigurationBuilder AddRewriteFile(this IConfigurationBuilder builder, string path, bool optional, bool reloadOnChange)
+        {
+            return AddRewriteFile(builder, provider: null, path: path, optional: optional, reloadOnChange: reloadOnChange);
+        }
+
+        public static IConfigurationBuilder AddRewriteFile(this IConfigurationBuilder builder, IFileProvider provider, string path, bool optional, bool reloadOnChange)
+        {
+
```
Comment (muratg) [inline samples/Rewrite/FileParser/RewriteConfigurationExtensions.cs]: nit: spaces

File: samples/Rewrite/FileParser/RewriteConfigurationFileParser.cs [hunk 6]
```diff
@@ -0,0 +1,38 @@
...
+{
+    public class RewriteConfigurationFileParser
+    {
+        private StreamReader _reader;
+        public void Parse(Stream input)
+        {
+            _reader = new StreamReader(input);
+            string line = null;
+            while ((line = _reader.ReadLine()) != null) {
+                List<string> tokens = RewriteTokenizer.TokenizeRule(line);
```
Comment (muratg) [inline samples/Rewrite/FileParser/RewriteConfigurationFileParser.cs]: nit: var

File: samples/Rewrite/FileParser/RewriteConfigurationFileParser.cs [hunk 7]
```diff
@@ -0,0 +1,38 @@
...
+                if (tokens.Count < 3)
+                {
+                    // I think this is a syntax error no matter what
```
Comment (muratg) [inline samples/Rewrite/FileParser/RewriteConfigurationFileParser.cs]: What will the behavior be here?

File: samples/Rewrite/FileParser/RewriteReaderException.cs [hunk 8]
```diff
@@ -0,0 +1,16 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading.Tasks;
+
+namespace Rewrite.FileParser
+{
+    public class RewriteReaderException : Exception
```
Comment (muratg) [inline samples/Rewrite/FileParser/RewriteReaderException.cs]: Do we need a custom Exception type? Does any one of the existing .NET exceptions make sense here instead?

File: samples/Rewrite/FileParser/RewriteTokenizer.cs [hunk 9]
```diff
@@ -0,0 +1,114 @@
...
+                    // Need to progress such that the next character is not evaluated.
+                    if (!context.Next())
+                    {
+                        // Means that a character was not escaped appropriately Ex: "foo\"
+                        throw new ArgumentException();
+                    }
+                }
+                else if (context.Current == Space)
+                {
+
```
Comment (muratg) [inline samples/Rewrite/FileParser/RewriteTokenizer.cs]: nit: space

File: samples/Rewrite/FileParser/RewriteTokenizer.cs [hunk 10]
```diff
@@ -0,0 +1,114 @@
...
+            public TokenContext(string line)
+            {
+                _template = line;
+                _index = -1;
+            }
+            public char Current
+            {
+                get { return (_index < _template.Length && _index >= 0) ? _template[_index] : (char)0; }
+            }
+            public string Error
```
Comment (muratg) [inline samples/Rewrite/FileParser/RewriteTokenizer.cs]: Single line unless you have any logic in get or set.

File: samples/Rewrite/Project_Readme.html [hunk 11]
```diff
@@ -0,0 +1,187 @@
+﻿<!DOCTYPE html>
```
Comment (Tratcher) [inline samples/Rewrite/Project_Readme.html]: delete

**Pull request closed without merging**